File: mobile/android/bench/app/src/main/java/com/siq/bench/metrics/MetricLogger.kt
package com.siq.bench.metrics

import android.util.Log
import com.siq.bench.RuntimeConfig
import kotlin.time.Duration

interface MetricLogger {
    fun logRun(
        config: RuntimeConfig,
        frameCount: Int,
        p50: Duration,
        p95: Duration,
        fps: Double,
        coldStart: Duration,
        memoryMb: Double,
        batteryDrainPercent: Double,
    )
}

class LogcatMetricLogger : MetricLogger {
    override fun logRun(
        config: RuntimeConfig,
        frameCount: Int,
        p50: Duration,
        p95: Duration,
        fps: Double,
        coldStart: Duration,
        memoryMb: Double,
        batteryDrainPercent: Double,
    ) {
        Log.i(
            "SIQBench",
            "${config.runtime}-${config.accelerator}: frames=$frameCount p50=${p50.inWholeMilliseconds}ms p95=${p95.inWholeMilliseconds}ms fps=${"%.1f".format(fps)} " +
                "coldStart=${coldStart.inWholeMilliseconds}ms mem=${"%.1f".format(memoryMb)}MB battery=${"%.2f".format(batteryDrainPercent)}%",
        )
    }
}

File: mobile/ios/bench/MetricLogger.swift
import Foundation

protocol MetricLogger {
    func log(config: RuntimeConfig, metrics: Metrics)
}

final class ConsoleMetricLogger: MetricLogger {
    func log(config: RuntimeConfig, metrics: Metrics) {
        print("[SIQBench] \(config.runtime)-\(config.accelerator) frames=\(metrics.frames) " +
              "p50=\(metrics.p50 * 1000)ms p95=\(metrics.p95 * 1000)ms fps=\(metrics.fps) " +
              "coldStart=\(metrics.coldStart * 1000)ms mem=\(metrics.peakMemory)MB drain=\(metrics.batteryDrain)%")
    }
}

func runBenchHarness(logger: MetricLogger = ConsoleMetricLogger(), frameSource: FrameSource = InMemoryFrameSource()) {
    let runner = BenchmarkRunner(logger: logger, frameSource: frameSource)
    runner.runSweep()
}
